[QUZ-69][FEATURE] 좋아요 구현

JiwonKKang opened the pull request:
<!-- PR의 제목은 "[Jira 티켓 코드] [종류(ex.FEATURE)] 이슈이름" 으로 작성 -->

## ✨ 구현 기능 명세

<!-- 해당 이슈에서 구현한 기능을 간단하게 작성 -->

- 동시성 이슈 redis를 통해 해결
- 검색 서비스와 동기화문제 역정규화와 스케쥴링을 통한 배치 업데이트로 해결
- 다중 서버에서 스케쥴링 shedlock으로 해결

## ✅ PR Point

<!-- 코드리뷰를 받고 싶은 부분, 새로운 지식을 얻게 된 부분 등등 공유하고 싶은 점을 작성 -->


## 😭 어려웠던 점

<!-- 해결하기 어려웠거나 시간이 오래 걸린 부분 작성 -->

와 진짜 좋아요 하나 구현하는데 이렇게 오래 걸릴줄 몰랐습니다....

동시성, 동기화, 다중서버 환경 고려할게 너무 많네요.... 일단 해당 이슈들 해결한다고 해결하긴했는데 최선의 방법인지는 확신이 없습니다..

좋아요 동시성 문제는 redis INCR을 통해 해결했고 사용자가 좋아요를 누르면 redis에 10초동안 쌓이다가 스케쥴링 처리로 퀴즈 테이블에 배치 업데이트가 됩니다!

퀴즈 테이블이 업데이트되면 퀴즈 테이블을 바라보고있던 debezium이 update 이벤트를 발행해서 해당 이벤트를 검색서비스에서 처리해 동기화하는 구조입니다!

다중서버에서 동기화 스케쥴링이 각각 돌아가면 문제가 될거라고 생각해 mysql shedlock을 통해 한번만 실행되도록 보장했습니다... 너무 힘드네요..

+) 민혁님의 피드백으로 좋아요를 한 사람들은 set을 이용해 구현했습니다. 하지만 set의 size로 좋아요수를 동기화하면 10초마다 기존 like_count에 + 연산을 하는 batch update와 함께 다시 0으로 초기화해야하는데 set_size로 좋아요 수를 판단하면 좋아요한 사람들을 캐시에서 모두 날려야하는 상황이 발생해서 `좋아요 한 사람`과 `좋아요 수`는 각각 set과 key를 이용해 관리하도록했습니다!

+) 아래는 제가 고민했던 내용입니다

과제1 : 동시성 이슈없이 좋아요수를 해결한다 -> redis
과제2: HTTP 요청으로 매번 받아오는 로직의 성능 문제 해결

과제 2번이 좀 큰 문제였는데 해결 방법이 생각났다.

먼저 좋아요 수 같은 경우에는 동시성 이슈가 발생할 가능성이 높고 클라이언트에게 실시간성이 보장될 필요는 크게 없다고 판단했다. 

따라서 redis(싱글 스레드)의 count를 통해 좋아요 수를 올리고 주기적으로 퀴즈 서비스 내의 quizzes 테이블에 update(redis <-> 퀴즈 서비스내의 DB 동기화)를 한다. 이렇게 update가 될때마다 debezium이 CDC를 통해 update 이벤트를 카프카에 전송하게되고 검색 서비스는 이 이벤트를 받아 퀴즈 도큐먼트의 좋아요 수를 업데이트한다.

댓글 수는 동시성 이슈 발생 가능성이 낮기때문에 따로 redis로 관리하지않고 사용가 댓글을 달때마다 read lock을 걸어서 quiz를 가져온뒤에 commentCount를 업데이트 해준다. 이 트랜잭션이 끝나면 마찬가지로 CDC에 의해 업데이트 이벤트가 전송되고 검색서비스는 이를 처리해 commentCount를 업데이트한다.

JiwonKKang committed:
feat : Add like API

- 동시성 이슈 redis를 통해 해결
- 검색 서비스와 동기화문제 역정규화와 스케쥴링을 통한 배치 업데이트로 해결
- 다중 서버에서 스케쥴링 shedlock으로 해결

JiwonKKang committed:
feat : Add check like status

NaMinhyeok commented:
이거는 근데 개인적인 생각인데 제가 이번에 게임도메인을 하면서 레디스에 대해서 좀 공부를 하게 되었는데요 지금 현재 구현하신 방식이 잘못됐다 이런건 아니고 이런 방식은 어떤지 제안해봅니다. 
이렇게 하는게 좀 더 우아하게(?) 레디스를 사용하는게 아닐까 라고 생각하기도 하고, 실제로 좋아요는 이러한 방식으로 많이 구현한다고 하네요.

일단 설명하기에 앞서서 레디스가 단순하게 key - value라는건 알고 계실건데 레디스에 value에 다양한 자료형이 존재한다고 해요. 그래서 좋아요를 구현하는 과정에 있어서는 value로 set을 이용하는게 어떨까라고 생각합니다. Set,집합 자료구조는 뭐 잘 아실거라고 생각합니다.

그래서 Quiz의 id를 key로 하고, quiz-like:{incr-key} 이런식으로 key를 세팅하고 value에는 set을 이용하셔서 userId를 집어넣으면 좋아요의 중복도 일어나지 않고, 좋아요의 갯수를 세어야 할 때 집합의 크기를 확인하여서 좋아요 갯수를 확인할 수 있습니다. Redis의 명령어로는 SCARD라고 하네요.

지금 현재 구조를 보면 quizId를 key로 하고 좋아요가 눌리면 id가 올라가고 id가 내려가도록 구현이 되어있는걸로 보입니다. 물론 이렇게 하는게 잘못된건 아닌데 실제로 동기화를 위해서 스케줄링도 짧게 두고 있어서 좋지않은 성능을 가질 수도있다고 생각하긴합니다. 그래서 그럼 어떻게 동기화할건데? 라고하면 저도 잘 모르겠지만...
개인적으로는 그냥 검색도메인도 좋아요수를 redis에서 가져오면 되는거아닌가? 라고 간단하게 생각해봅니다.. 아니면 그냥 RDB로 바로저장하지 않고 좋아요수를 레디스에서 바로 동기화하는방법?도 불가능한건 아니지않을까 라고 생각은 해봤습니다. 구현이 쉽지는 않겠지만요

사실은 CDC나 다른 서버와의 동기화 문제도 있기 때문에 단순히 생각할 문제는 아니라고 생각하기는 하는데 레디스를 좀 더 레디스답게 써보면 좋을 것 같아서 한번 추천해봅니다 !

HMWG commented on quiz-service/quiz-infra/src/main/kotlin/com/grepp/quizy/quiz/infra/quiz/repository/QuizJpaBatchRepositoryImpl.kt:
JDBC로 batch update하는 방법만 알고 있었는데 이렇게 entityManager에서 batch update 하는 방법도 있네요
잘 배워갑니다~

HMWG reviewed (approved):
동시성 이슈와 관련해서 다른 서버에도 같이 처리해줘야 하니 복잡하네요

mySQL에 스케줄링을 통해 batch update가 발생할 때 마다 cdc를 통해 카프카 이벤트가 발생한다라고 이해했는데 맞는지 모르곘습니다

여하튼 다중 서버이다 보니 동기화에서도 생각할게 많네요 고생 많으셨습니다..!

JiwonKKang committed:
refactor : Change to Redis set

- 좋아요한 사람 redis set을 통해 관리
- 좋아요 수는 10초마다 0으로 초기화되며 동기화해야하기때문에 기존 방법 유지
- 배치 업데이트 하는동안에 오른 좋아요 수 체크

JiwonKKang committed:
fix : Fix redis key race condition

- 키값을 가져오고 해당 키를 삭제하는 사이에 좋아요 ㅅ가 변경되면 해당 좋아요 수는 유실되는 문제를 해결

s0o0bn commented:
고생하셨습니다...!! 

MSA에 서버 확장도 고려하다보니 진짜 좋아요 하나도 엄청 복잡해지네요...
